anonlink/serialization.py

```diff
@@ -0,0 +1,221 @@
+import array as _array
+import heapq as _heapq
+import io as _io
+import struct as _struct
+import typing as _typing
+
+from . import typechecking as _typechecking
+
+# https://docs.python.org/3/library/struct.html#format-characters
+_STRUCT_UINT_LEN_TO_FMT = {1: 'B', 2: 'H', 4: 'L', 8: 'Q'}
+_STRUCT_FLOAT_LEN_TO_FMT = {2: 'e', 4: 'f', 8: 'd'}
+
+# https://docs.python.org/3/library/array.html#module-array
+_ARRAY_UINT_TYPES = 'BIHLQ'
+_ARRAY_FLOAT_TYPES = 'df'
+_ARRAY_UINT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                          for t in _ARRAY_UINT_TYPES}
+_ARRAY_FLOAT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                           for t in _ARRAY_FLOAT_TYPES}
+
+_HEADER_STRUCT = _struct.Struct('<BBBB')
+
+
+_CandidatePair = _typing.Tuple[float, int, int, int, int]
+_CandidatePairIter = _typing.Iterable[_CandidatePair]
+
+# This is subject to change.
+# The format is composed of a header and a sequence of entries.
+# The header is composed of:  version (1 byte), number of bytes in
+# similarity (1 byte), number of bytes in dataset index (1 byte), number
```

Comment:
"number of bytes in similarity" isn't clear until reading the next section.

Reply:
I’ve (hopefully) made it more clear.